sql/analyzer/resolve_ctes.go

```diff
@@ -72,26 +72,28 @@ func resolveCtesInNode(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope,
 	// second pass, that reference gets resolved to the subquery alias of cte1's definition. Then we're done.
 	// We iterate until the tree stops changing, or until we hit our limit.
 	var cur, prev sql.Node
+	mod := true
 	cur = n
-	for i := 0; i < maxCteDepth && !nodesEqual(prev, cur); i++ {
+	for i := 0; i < maxCteDepth && mod; i++ {
```

Comment:
Not sure about this, it makes it possible for an analyzer rule to misbehave and claim it didn't change anything when it actually did.

On the other hand, not calling DeepEquals on every node is probably a huge performance savings.

I think I would be more comfortable getting rid of DeepEquals here if a rule had to opt-in to saying it changed nothing, rather than opt-in to saying something changed.